src/sinks/socket.rs

```diff
@@ -221,37 +226,45 @@ mod test {
                 ..Default::default()
             },
         });
-        let stream = MaybeTlsSettings::from_config(&config, true).unwrap();
 
         // Only accept two connections.
-        let fut = stream
-            .bind(&addr)
-            .unwrap()
-            .incoming()
-            .take(2)
-            .for_each(move |mut socket| {
-                conn_counter1.fetch_add(1, Ordering::SeqCst);
-                loop {
-                    if let Ok(Async::Ready(_)) = close_rx.poll() {
-                        socket.get_ref().unwrap().shutdown(Shutdown::Write).unwrap();
-                    }
-
-                    let mut buf = vec![0u8; 11];
-                    match socket.read(&mut buf) {
-                        Err(error) if error.kind() == ErrorKind::WouldBlock => {}
-                        Err(error) => panic!("{}", error),
-                        Ok(n) => {
-                            if n == 0 {
+        let jh = rt.spawn_handle_std(async move {
+            let tls = MaybeTlsSettings::from_config(&config, true).unwrap();
+            let mut listener = tls.bind(&addr).await.unwrap();
+            listener
+                .incoming()
+                .take(2)
+                .for_each(|connection| {
+                    let close_rx = Arc::clone(&close_rx);
+
+                    conn_counter1.fetch_add(1, Ordering::SeqCst);
+                    let msg_counter1 = Arc::clone(&msg_counter1);
+
+                    async move {
+                        let mut stream: MaybeTlsIncomingStream<TcpStream> = connection.unwrap();
+                        let mut cx = Context::from_waker(&noop_waker_ref());
+                        loop {
+                            if close_rx.lock().await.try_recv().is_ok() {
+                                stream.get_ref().unwrap().shutdown(Shutdown::Write).unwrap();
                                 break;
-                            } else {
-                                msg_counter1.fetch_add(1, Ordering::SeqCst);
+                            }
+
+                            match Pin::new(&mut stream).poll_read(&mut cx, &mut [0u8; 11]) {
```

Comment:
This seems to have gotten a lot more complicated because you switched from `read` to `poll_read`. Is that something we need to do? I belive the initial logic was that this socket will already be in non-blocking mode and a normal read works just fine.

Reply:
[TcpStream](https://docs.rs/tokio/0.2.22/tokio/net/struct.TcpStream.html) from new tokio not implement [std::io::Read](https://doc.rust-lang.org/std/io/trait.Read.html), I think this particular example can be simplified, I'll try.